File: fix_enrich.php
#!/usr/bin/env php
<?php
/**
 * Fix Enrich - 批量修复补全文件中问题行的脚本（PHP版）
 *
 * 扫描已有的 _补全.xlsx 文件，根据预定义规则自动识别有问题的行，
 * 重新调用 enrichProduct() 进行补全，覆盖原数据。
 *
 * 检查规则:
 *   1. 铁线莲三级必须是「木本藤本」（不是草本藤本）
 *   2. 绣球四级必须是「绣球花科」（不是绣球科）
 *   3. 枫树四级必须是「槭树科」（不是槒树科）
 *   4. 百子莲三级必须是「鳞茎」（不是多年生草本）
 *   5. 品牌HY/HY-/Encore Azaleas → 统一为「安酷」
 *   6. 一级~四级分类/品牌完全缺失 → 重新补全
 *
 * 使用方法:
 *   php fix_enrich.php
 */

define('SOURCE_FILE', '1年内销售库存大于10育种+上海_补全.xlsx');
define('CATEGORY_FILE', '商品分类目录.xlsx');

// ===========================================================================
// 工具函数（与 batch_enrich.php 共享）
// ===========================================================================

function loadCategories(string $path): array
{
    if (!class_exists('\PhpOffice\PhpSpreadsheet\IOFactory')) {
        die("错误: 需要安装 phpoffice/phpspreadsheet\n");
    }
    $spreadsheet = \PhpOffice\PhpSpreadsheet\IOFactory::load($path);
    $sheet = $spreadsheet->getActiveSheet();
    $rows = $sheet->toArray();

    $categories = [];
    for ($i = 1; $i < count($rows); $i++) {
        $row = $rows[$i];
        if (!empty($row[0])) {
            $categories[] = [
                '一级分类' => $row[0],
                '二级分类' => $row[1] ?? '',
                '三级分类' => $row[2] ?? '',
                '四级分类' => $row[3] ?? '',
            ];
        }
    }
    return $categories;
}

function loadExcel(string $path): array
{
    if (!class_exists('\PhpOffice\PhpSpreadsheet\IOFactory')) {
        die("错误: 需要安装 phpoffice/phpspreadsheet\n");
    }
    $spreadsheet = \PhpOffice\PhpSpreadsheet\IOFactory::load($path);
    return $spreadsheet->getActiveSheet()->toArray();
}

function saveExcelWithPhpSpreadsheet(array $data, string $path): void
{
    $spreadsheet = new \PhpOffice\PhpSpreadsheet\Spreadsheet();
    $sheet = $spreadsheet->getActiveSheet();

    foreach ($data as $rowIdx => $row) {
        foreach ($row as $colIdx => $value) {
            $sheet->setCellValueByColumnAndRow($colIdx + 1, $rowIdx + 1, $value ?? '');
        }
    }

    $writer = \PhpOffice\PhpSpreadsheet\IOFactory::createWriter($spreadsheet, 'Xlsx');
    $writer->save($path);
}

function findFile(string $filename): ?string
{
    $scriptDir = __DIR__;
    $home = getenv('HOME') ?: exec('echo $HOME');

    $candidates = [
        $scriptDir . '/' . $filename,
        $home . '/Desktop/沈飞mac/skills/shopping/' . $filename,
    ];

    foreach ($candidates as $p) {
        if (file_exists($p)) return $p;
    }
    return null;
}

function autoload(): void
{
    $paths = [
        __DIR__ . '/vendor/autoload.php',
        dirname(__DIR__) . '/vendor/autoload.php',
    ];
    foreach ($paths as $p) {
        if (file_exists($p)) {
            require_once $p;
            return;
        }
    }
}

// ===========================================================================
// 问题行检查规则
// ===========================================================================

/**
 * 检查指定行是否有已知问题
 *
 * @param int    $rowIdx  Excel行号（从2开始，数据行）
 * @param string $name    商品名
 * @param string $yiji    一级分类
 * @param string $erji    二级分类
 * @param string $sanji   三级分类
 * @param string $siji    四级分类
 * @param string $brand   品牌
 * @param string $variety 品种
 *
 * @return array [isProblematic: bool, reason: string]
 */
function checkRow(int $rowIdx, string $name, string $yiji, string $erji, string $sanji, string $siji, string $brand, string $variety): array
{
    // ---- 规则1: 铁线莲三级分类 ----
    // 铁线莲是木本藤本，不是草本藤本
    if (mb_strpos($name, '铁线莲') !== false && $sanji === '草本藤本') {
        return [true, '铁线莲三级分类错误（草本藤本→木本藤本）'];
    }

    // ---- 规则2: 绣球四级分类 ----
    // 绣球的科名是「绣球花科」，不是「绣球科」
    if ($siji === '绣球科') {
        return [true, "绣球四级分类错误（{$siji}→绣球花科）"];
    }

    // ---- 规则3: 枫树四级分类 ----
    // 枫树的科名是「槭树科」，不是「槒树科」
    if ($siji === '槒树科') {
        return [true, "枫树四级分类错误（{$siji}→槭树科）"];
    }

    // ---- 规则4: 百子莲三级分类 ----
    // 百子莲是鳞茎植物，不是多年生草本
    if (mb_strpos($name, '百子莲') !== false && $sanji === '多年生草本') {
        return [true, '百子莲三级分类错误（多年生草本→鳞茎）'];
    }

    // ---- 规则5: 安酷杜鹃品牌名称归一 ----
    // Encore Azaleas / HY / HY- 统一输出为「安酷」
    $brandTrim = trim($brand);
    if ($brandTrim && (
        in_array($brandTrim, ['HY', 'Encore Azaleas（安酷杜鹃）', 'Encore Azaleas'], true) ||
        strpos($brandTrim, 'HY') === 0
    )) {
        return [true, "品牌写法错误（{$brand}→安酷）"];
    }

    // ---- 规则6: 数据完全缺失 ----
    // 一级分类为空说明该行从未成功补全
    if ($yiji === null || $yiji === '') {
        return [true, '数据完全缺失'];
    }

    // 无问题
    return [false, ''];
}

// ===========================================================================
// 主流程
// ===========================================================================

function main(): void
{
    autoload();

    // ---- 1. 定位补全文件 ----
    $excelPath = findFile(SOURCE_FILE);
    if (!$excelPath) {
        echo "文件不存在: " . SOURCE_FILE . "\n";
        exit(1);
    }

    echo "读取文件: $excelPath\n";
    $allRows = loadExcel($excelPath);
    $totalRows = count($allRows) - 1;
    echo "总行数: $totalRows\n\n";

    // ---- 2. 加载分类目录 ----
    $catPath = findFile(CATEGORY_FILE);
    if (!$catPath) {
        echo "错误: 找不到 " . CATEGORY_FILE . "\n";
        exit(1);
    }

    echo "加载分类目录: $catPath\n";
    $categories = loadCategories($catPath);
    echo "共 " . count($categories) . " 条分类记录\n\n";

    // ---- 3. 全量扫描，收集问题行 ----
    echo "=== 检查所有行 ===\n";
    $problematic = [];

    for ($rowIdx = 2; $rowIdx <= count($allRows); $rowIdx++) {
        $row = $allRows[$rowIdx - 1] ?? [];
        $goodsName = $row[1] ?? '';   // B列: goods_name
        $yiji     = $row[7] ?? '';   // H列: 一级分类
        $erji     = $row[8] ?? '';   // I列: 二级分类
        $sanji    = $row[9] ?? '';   // J列: 三级分类
        $siji     = $row[10] ?? '';  // K列: 四级分类
        $brand    = $row[11] ?? '';  // L列: 品牌
        $variety  = $row[13] ?? '';  // N列: 品种

        [$isBad, $reason] = checkRow($rowIdx, $goodsName, $yiji, $erji, $sanji, $siji, $brand, $variety);

        if ($isBad) {
            $problematic[] = [
                'row' => $rowIdx,
                'name' => $goodsName,
                'reason' => $reason,
                '当前' => "三级={$sanji}, 四级={$siji}, 品牌={$brand}",
            ];
        }
    }

    echo "发现问题行: " . count($problematic) . " 条\n\n";
    foreach ($problematic as $p) {
        echo "  Row {$p['row']} [" . mb_substr($p['name'], 0, 30) . "]\n";
        echo "    原因: {$p['reason']}\n";
        echo "    当前: {$p['当前']}\n\n";
    }

    if (empty($problematic)) {
        echo "没有发现问题行，无需修复\n";
        exit(0);
    }

    // ---- 4. 逐行重新补全 ----
    echo "开始重新补全 " . count($problematic) . " 条...\n\n";
    $fixedCount = 0;
    $errorCount = 0;
    $total = count($problematic);

    foreach ($problematic as $idx => $item) {
        $rowIdx = $item['row'];
        $goodsName = $item['name'];

        // 进度条
        $current = $idx + 1;
        $pct = $current / $total * 100;
        $barLen = 30;
        $filled = (int)($barLen * $current / $total);
        $bar = str_repeat('█', $filled) . str_repeat('░', $barLen - $filled);
        $extra = mb_substr($goodsName, 0, 25);
        echo "\r[{$bar}] " . sprintf("%.1f%% (%d/%d) - %s", $pct, $current, $total, $extra);
        echo str_repeat(' ', 20);

        $enriched = enrichProduct($goodsName, '', $categories);

        if (isset($enriched['error'])) {
            $errorCount++;
            echo "\n  错误: {$enriched['error']}\n";
        } else {
            $fixedCount++;
            // 更新数据行中的8个补全列（第8列N~第15列O）
            $enrichCols = ['一级分类', '二级分类', '三级分类', '四级分类', '品牌', '品类', '品种', '百科信息'];
            $rowNum = $rowIdx; // 行号从1开始，$rowIdx就是数据行号

            foreach ($enrichCols as $colIdx => $colName) {
                // 列号: 一级分类=H(8), 二级分类=I(9), ... 百科信息=O(15)
                $colLetter = chr(ord('H') + $colIdx); // H=8, I=9, ...
                $allRows[$rowNum - 1][7 + $colIdx] = $enriched[$colName] ?? '';
            }
        }

        // 每10行保存一次
        if ($current % 10 === 0) {
            saveExcelWithPhpSpreadsheet($allRows, $excelPath);
        }
    }

    echo "\n\n修复完成！成功: $fixedCount, 失败: $errorCount\n";
    saveExcelWithPhpSpreadsheet($allRows, $excelPath);
    echo "已保存: $excelPath\n";
}

// ===========================================================================
// 入口
// ===========================================================================
main();

File: fix_enrich.py
#!/usr/bin/env python3
"""
Fix Enrich - 批量修复补全文件中问题行的脚本

扫描已有的 _补全.xlsx 文件，根据预定义规则自动识别有问题的行，
重新调用 enrich_product() 进行补全，覆盖原数据。

检查规则（每次运行会根据实际情况更新）:
  1. 铁线莲三级必须是「木本藤本」（不是草本藤本）
  2. 绣球四级必须是「绣球花科」（不是绣球科）
  3. 枫树四级必须是「槭树科」（不是槒树科）
  4. 百子莲三级必须是「鳞茎」（不是多年生草本）
  5. 品牌HY/HY-/Encore Azaleas → 统一为「安酷」
  6. 一级~四级分类/品牌完全缺失 → 重新补全

使用方法:
    python3 fix_enrich.py

说明:
    - 仅处理问题行，非问题行不会被重新调用
    - 每10行保存一次进度
    - 无问题行时直接退出
"""

import sys
import os
import json
import re
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from product_enricher import load_categories, enrich_product

import openpyxl


# ===========================================================================
# 问题行检查规则
# ===========================================================================

def check_row(row_idx, name, yiji, erji, sanji, siji, brand, variety):
    """
    检查指定行的分类数据是否有已知问题。

    参数:
        row_idx ~ variety: Excel中各列的原始值

    返回:
        (is_problematic: bool, reason: str)
        - True + 说明文字 表示需要重新补全
        - False + '' 表示数据正常
    """
    name = name or ''

    # ---- 规则1: 铁线莲三级分类 ----
    # 铁线莲是木本藤本，不是草本藤本
    if '铁线莲' in name and sanji == '草本藤本':
        return True, '铁线莲三级分类错误（草本藤本→木本藤本）'

    # ---- 规则2: 绣球四级分类 ----
    # 绣球的科名是「绣球花科」，不是「绣球科」
    if siji in ('绣球科',):
        return True, f'绣球四级分类错误（{siji}→绣球花科）'

    # ---- 规则3: 枫树四级分类 ----
    # 枫树的科名是「槭树科」，不是「槒树科」
    if siji in ('槒树科',):
        return True, f'枫树四级分类错误（{siji}→槭树科）'

    # ---- 规则4: 百子莲三级分类 ----
    # 百子莲是鳞茎植物，不是多年生草本
    if '百子莲' in name and sanji == '多年生草本':
        return True, '百子莲三级分类错误（多年生草本→鳞茎）'

    # ---- 规则5: 安酷杜鹃品牌名称归一 ----
    # Encore Azaleas / HY / HY- 统一输出为「安酷」
    if brand and (brand.strip() in ('HY', 'Encore Azaleas（安酷杜鹃）', 'Encore Azaleas') or brand.strip().startswith('HY')):
        return True, f'品牌写法错误（{brand}→安酷）'

    # ---- 规则6: 数据完全缺失 ----
    # 一级分类为空说明该行从未成功补全
    if yiji is None:
        return True, '数据完全缺失'

    # 无问题
    return False, ''


# ===========================================================================
# 主流程
# ===========================================================================

def main():
    # ---- 1. 定位补全文件 ----
    excel_path = Path(__file__).parent / '1年内销售库存大于10育种+上海_补全.xlsx'
    if not excel_path.exists():
        print(f"文件不存在: {excel_path}")
        sys.exit(1)

    print(f"读取文件: {excel_path}")
    wb = openpyxl.load_workbook(excel_path)
    ws = wb.active

    total_rows = ws.max_row - 1
    print(f"总行数: {total_rows}\n")

    # ---- 2. 加载分类目录 ----
    categories = None
    for parent in [excel_path.parent, Path(__file__).parent]:
        xlsx = parent / '商品分类目录.xlsx'
        if xlsx.exists():
            print(f"加载分类目录: {xlsx}")
            categories = load_categories(str(xlsx))
            print(f"共 {len(categories)} 条分类记录\n")
            break

    if not categories:
        print("错误: 找不到商品分类目录.xlsx")
        sys.exit(1)

    # ---- 3. 全量扫描，收集问题行 ----
    print("=== 检查所有行 ===")
    problematic = []
    for row_idx in range(2, ws.max_row + 1):
        goods_name = ws.cell(row_idx, 2).value
        yiji  = ws.cell(row_idx, 8).value   # 一级分类
        erji  = ws.cell(row_idx, 9).value   # 二级分类
        sanji = ws.cell(row_idx, 10).value  # 三级分类
        siji  = ws.cell(row_idx, 11).value  # 四级分类
        brand = ws.cell(row_idx, 12).value  # 品牌
        variety = ws.cell(row_idx, 14).value  # 品种

        is_bad, reason = check_row(row_idx, goods_name, yiji, erji, sanji, siji, brand, variety)
        if is_bad:
            problematic.append({
                'row': row_idx,
                'name': goods_name,
                'reason': reason,
                '当前': f'三级={sanji}, 四级={siji}, 品牌={brand}'
            })

    print(f"发现问题行: {len(problematic)} 条\n")
    for p in problematic:
        print(f"  Row {p['row']} [{str(p['name'])[:30]}]")
        print(f"    原因: {p['reason']}")
        print(f"    当前: {p['当前']}\n")

    if not problematic:
        print("没有发现问题行，无需修复")
        sys.exit(0)

    # ---- 4. 逐行重新补全 ----
    print(f"开始重新补全 {len(problematic)} 条...\n")
    fixed_count = 0
    error_count = 0

    for i, item in enumerate(problematic):
        row_idx = item['row']
        goods_name = str(item['name'])

        # 进度条
        current = i + 1
        pct = current / len(problematic) * 100
        bar_len = 30
        filled = int(bar_len * current / len(problematic))
        bar = '█' * filled + '░' * (bar_len - filled)
        print(f"\r[{bar}] {pct:.1f}% ({current}/{len(problematic)}) - {goods_name[:25]}", end='', flush=True)

        result = enrich_product(goods_name, '', categories)

        if 'error' in result:
            error_count += 1
            print(f"\n  错误: {result['error']}")
        else:
            fixed_count += 1
            # 写入8个补全字段（第8列起）
            enrich_cols = ['一级分类', '二级分类', '三级分类', '四级分类', '品牌', '品类', '品种', '百科信息']
            for j, col_name in enumerate(enrich_cols):
                ws.cell(row_idx, 8 + j).value = result.get(col_name, '')

        # 每10行保存一次，防止意外中断丢失
        if (i + 1) % 10 == 0:
            wb.save(excel_path)

    print(f"\n\n修复完成！成功: {fixed_count}, 失败: {error_count}")
    wb.save(excel_path)
    print(f"已保存: {excel_path}")


if __name__ == '__main__':
    main()